File: education/academic_management/page/student_module_asses/student-module-asses.py
import frappe
from frappe import _

@frappe.whitelist()
def get_ca_report(module_enrollment):

    values = frappe.get_value(
        "Module Enrolment Key",
        module_enrollment,
        ["college", "programme", "academic_term", "module"],
        as_dict=True
    )

    if not values:
        frappe.throw(_("Module Enrolment Key not found"))

    # =========================
    # COMPONENTS
    # =========================
    components = frappe.db.sql("""
        SELECT 
            mai.assessment_name
        FROM `tabModule Assessment Item` mai
        INNER JOIN `tabModule Assessment Criteria` mac 
            ON mai.parent = mac.name
        WHERE 
            mac.college = %s
            AND mac.academic_term = %s
            AND mac.programme = %s
            AND mac.module = %s
            AND mai.docstatus=1
        ORDER BY mai.idx
    """, (values.college, values.academic_term, values.programme, values.module), as_dict=True)

    # frappe.throw(str(components))

    # =========================
    # RAW DATA
    # =========================
    rows = frappe.db.sql("""
        SELECT 
            al.student,
            al.student_name,
            al.assessment_component,
            al.weightage_achieved
        FROM `tabAssessment Ledger` al
        WHERE 
            al.college = %(college)s
            AND al.academic_term = %(academic_term)s
            AND al.module = %(module)s
            AND al.programme = %(programme)s
            AND al.is_cancelled = 0
            And al.has_reassessment = 0
    """, values, as_dict=True)

    # =========================
    # STEP 1: AGGREGATE PER STUDENT + COMPONENT
    # =========================
    temp = {}

    for r in rows:
        key = (r.student, r.assessment_component)

        if key not in temp:
            temp[key] = {
                "student": r.student,
                "student_name": r.student_name,
                "component": r.assessment_component,
                "value": 0
            }

        temp[key]["value"] += r.weightage_achieved or 0

    # =========================
    # STEP 2: BUILD FINAL STRUCTURE
    # =========================
    students = {}

    for t in temp.values():
        sid = t["student"]

        if sid not in students:
            students[sid] = {
                "student": sid,
                "student_name": t["student_name"],
                "assessments": {},
                "total": 0
            }

        students[sid]["assessments"][t["component"]] = t["value"]
        students[sid]["total"] += t["value"]


    return {
        "components": components,
        "students": list(students.values())
    }

File: education/academic_management/page/student_module_asses/student_module_asses.js
frappe.pages['student-module-asses'].on_page_load = function(wrapper) {

	let page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Assessment Component Report',
		single_column: true
	});

	// ============================
	// FILTER UI
	// ============================
	let fieldContainer = $(`
		<div class="row" style="margin-bottom:15px;">
			<div class="col-md-4" id="module-enroll-field"></div>
			<div class="col-md-2" id="btn-field"></div>
		</div>
	`).appendTo(page.body);

	let module_enroll_key = frappe.ui.form.make_control({
		parent: $('#module-enroll-field'),
		df: {
			label: 'Module Enrollment',
			fieldname: 'module_enrollment',
			fieldtype: 'Link',
			options: 'Module Enrolment Key',
			reqd: 1
		},
		render_input: true
	});
	module_enroll_key.refresh();

	let btn = $(`
		<button class="btn btn-primary" style="margin-top:25px;">
			Load Report
		</button>
	`).appendTo('#btn-field');

	let dataContainer = $('<div style="margin-top:20px;"></div>').appendTo(page.body);

	// ============================
	// LOAD REPORT
	// ============================
	function load_report() {

		let module_enrollment = module_enroll_key.get_value();

		if (!module_enrollment) {
			frappe.msgprint("Please select Module Enrollment");
			return;
		}

		dataContainer.html("<p>Loading...</p>");

		frappe.call({
			method: "education.academic_management.page.student_module_asses.student-module-asses.get_ca_report",
			args: {
				module_enrollment: module_enrollment
			},
			callback: function(r) {

				let res = r.message;

				if (!res || !res.students || !res.students.length) {
					dataContainer.html("<b>No data found</b>");
					return;
				}

				let students = res.students;
				let components = res.components;

				// 🔥 extract clean component names
				let comp_names = components.map(c => c.assessment_name);

				// ============================
				// TABLE START
				// ============================
				let html = `
				<div style="overflow-x:auto;">
				<table class="table table-bordered table-sm" style="table-layout: fixed; width: 100%;">
				`;

				// ============================
				// HEADER
				// ============================
				// html += `<thead><tr>`;
				html += `
					<thead style="
						background: #f5f5f5;
						box-shadow: 0 2px 4px rgba(0,0,0,0.08);
						position: sticky;
						top: 0;
						z-index: 2;
					">
					<tr>
					`;

				html += `<th style="width:120px;">Student ID</th>`;
				html += `<th style="width:200px;">Student Name</th>`;

				comp_names.forEach(c => {
					html += `
						<th style="text-align:center; white-space:nowrap;">
							${c}
						</th>`;
				});

				html += `<th style="width:100px; text-align:center;">Total</th>`;
				html += `</tr></thead><tbody>`;

				// ============================
				// BODY
				// ============================
				students.forEach(student => {

					html += `<tr>`;

					// ID
					html += `<td>${student.student}</td>`;

					// NAME (from backend)
					html += `<td style="font-weight:500;">
						${student.student_name || ''}
					</td>`;

					// COMPONENT VALUES
					comp_names.forEach(comp => {
						let val = student.assessments[comp] || 0;
						html += `<td style="text-align:center;">${val}</td>`;
					});

					// TOTAL (safe + no float issues)
					html += `<td style="text-align:center; font-weight:bold; background:#f5f5f5;">
						${Number(student.total || 0)}
					</td>`;

					html += `</tr>`;
				});

				html += `</tbody></table></div>`;

				dataContainer.html(html);
			}
		});
	}

	btn.on("click", load_report);
};